rfcs/006 - Sync changes across databases.md

```diff
@@ -0,0 +1,84 @@
+
+## Alternatives Considered
+
+There are several types of solutions to replicate staging infrastructure changes in production.
+
+### 1. Dump a database and import the dump into another database
+
+An `export` route in the engine implemented using `pgdump` with logic to attempt to make it generic like excluding `aws_account` from the dump and so on. The dump is then intended to be passed into an `import` route for another database. `import` was deprecated for the time being due to the sheer amount of special-casing, which compounded per module, that had to be built within the functionality of export and import.
+
+### 2. Sync a database to a different cloud account than the one it is connected to
+
+An IaSQL function that takes the alias of another dataabse owned by the same user as input. `iasql_import(‘staging’)` will return the same data structure as `iasql_commit` or `iasql_preview`. This function can reuse parts of the mappers and existing IaSQL functions to run diffs across two different databases or accounts. There could have an optional `dry_run` parameter or another function to preview the changes: `iasql_preview_import(‘staging’)` or `iasql_import(‘staging’, true)`.
+
+The steps to replicate recent database/account A changes in database/account B when calling `iasql_import('A')` from within database B:
+1) Regular sync database A to cloud account A
+2) Regular sync database B to cloud account B
+3) New import sync functionality where database B is pointed to cloud account A directly
+
+The new import sync functionality described in #3 would expand the mapper definitions to include a new `import` function and laxer equality function for each module.
+
+### 3. Interactive `git rebase` style sync
+
+Similar to option 2, but when conflict arises prompt the user to handle the merge conflicts between the two cloud accounts or databases interactively. The PostgreSQL REPL doesn't allow interactive input so that would have to happen using a two-phase process via an interim table or the dashboard. This option requires a lot of resources to implement out of the gate. The current proposal and if this workflow proves to be primordial to users we can eventually prioritize it.
+
+### 4. Use existing PostgreSQL data synchronization tools
+
+There is a breadth of tools to keep two databases in sync. However, they are not enough to satisfy our requirements for this problem because the data should not be copied over exactly as-is. We have cloud-generated IDs as columns in a lot of tables that are specific to a resource in each account like an EC2 instance id.
+
+### 5. Emulate a declaration via ORMs with introspection + Upserts
```

Comment:
I was thinking that we could just do a similar procedure as we do now with the engine with sync and commit, but using two accounts:
1. Consider database on _account 1_ as the source of truth
2. Consider cloud on _account 2_ as the cloud that needs to be updated
3. With the special consideration about different aws_accounts, regions, etc... let the engine run its logic, and perform changes to reflect status of _account 1_ into _account 2_.

Reply:
I am not sure if this comment was from after or before the sprint planning meeting. @dfellis point to this was that every module will have different edge cases outside of just the different `@cloudId`s in _account 1_ and _account 2_ respectively. If you also factor likely differences in the capacity of resources and values of secrets across staging and production, since different teams setup staging and prod differently, it would make this process extremely error-prone.

Reply:
> I was thinking that we could just do a similar procedure as we do now with the engine with sync and commit, but using two accounts:
> 
> 1. Consider database on _account 1_ as the source of truth
> 2. Consider cloud on _account 2_ as the cloud that needs to be updated
> 3. With the special consideration about different aws_accounts, regions, etc... let the engine run its logic, and perform changes to reflect status of _account 1_ into _account 2_.

btw this alternative is basically #2 in the RFC, but lmk if I am misinterpreting something and I can list it as a separate alternative